feat(icons): filetype detection via shebang

nobe4 opened the pull request:
This PR adds a step in the columns making: read the first line of the file and use that to compute a "fallback" filetype in case the filename alone cannot be used to determine the filetype.

This is then passed to the `IconProvider` which uses it if the icon is nil, or default.

This works well in initial tests, but `vim.fn.readfile(name, '', 1)` only read files in the current folder, I need to understand how I can make that work when changing folders. However, once this is done, I think this will provide some nice additional icons.

Looking forward to getting a review/suggestions :)

nobe4 committed:
test(files): try to load the ft from the first line of the file

nobe4 committed:
feat(utils): make the icon provider default to the specified filetype

stevearc commented on lua/oil/util.lua:
```diff
@@ -930,19 +930,26 @@ M.get_icon_provider = function()
-      return mini_icons.get(type == "directory" and "directory" or "file", name)
+    return function(type, name, ft)
+      local icon, hl, is_default = mini_icons.get(type == "directory" and "directory" or "file", name)
+      if icon == nil or is_default then
+        icon, hl = mini_icons.get("filetype", ft or "")
+      end
```
Probably should only do this when `ft` is non-nil, right?

stevearc replied:
Actually, I think for both devicons and MiniIcons it makes sense to use the filetype _first_ if it's provided, as that implementation should both be faster and more correct, and only use the original logic if the filetype is nil.

nobe4 replied:
I have pushed some commits to address this, mini.icons changed, but devicons actually was simplified, since it handles the `ft == nil` logic already.

stevearc reviewed (changes requested):
I mentioned this [in the issue](https://github.com/stevearc/oil.nvim/issues/617#issuecomment-2927654702), but performance is a concern here. Passing a max lines of `1` to `readfile` is helpful, but this will still slow us down in large directories because we'll have to open a _lot_ of files. Could we maybe put this behind a config option so the default behavior is still fast?

nobe4 committed:
refactor(util): get icon, hl from ft if specified

As @stevearc mentioned, it would be better to use the ft directly if the
function gets it, and fallback to the original logic otherwise.

cc https://github.com/stevearc/oil.nvim/pull/618#discussion_r2119421575

nobe4 committed:
refactor(util): simplify the devicons logic

Upon looking at devicons' code, it seems that passing `ft` even when nil
will result in the same execution, and doing otherwise won't create any
performance improvements.

nobe4 committed:
feat(config): add key to toggle ft detection

As @stevearc suggested, it would be better to keep the default behavior
for performance reasons. Adding a config option to enable the ft
detection ensures this.

cc https://github.com/stevearc/oil.nvim/pull/618#pullrequestreview-2885911953

nobe4 committed:
docs(config): add field definition

nobe4 commented on lua/oil/columns.lua:
```diff
@@ -168,7 +168,14 @@ if icon_provider then
+
+      local ft = ""
+      if field_type == "file" then
+        local firstline = vim.fn.readfile(name, '', 1)
```
@stevearc, I have searched, and failed to find how to get absolute filepaths, or the current directory in the context of this function. Do you know how I can get it?

stevearc replied:
The render function takes a bufnr as the third argument
https://github.com/stevearc/oil.nvim/blob/685cdb4ffa74473d75a1b97451f8654ceeab0f4a/lua/oil/columns.lua#L15
You could parse the buffer name to get the parent dir like we do here
https://github.com/stevearc/oil.nvim/blob/685cdb4ffa74473d75a1b97451f8654ceeab0f4a/lua/oil/adapters/files.lua#L506-L507

nobe4 replied:
Awesome! Thanks for this!

Done in https://github.com/stevearc/oil.nvim/pull/618/commits/e2ad451f57c5cfa82ef7bded0b44f473e98a2919

nobe4 committed:
refactor(columns): leave ft to nil

The ft is compared against `nil` in the rest of the code, so don't set
it up to an empty string if missing.

nobe4 committed:
refactor(util): improve devicons' usage

`get_icon` favor the filename instead of the `ft`, so we need to use
`get_icon_by_filetype` to explicitly make the ft a priority.

This commits leaves the normal logic as a fallback if the icon is not
found with the filetype.

stevearc commented on lua/oil/config.lua:
```diff
@@ -100,6 +100,9 @@ local default_config = {
       { "type", "asc" },
       { "name", "asc" },
     },
+    -- Read the first line of the file and use vim's matching ft.
+    -- Can slow down the execution considerably for large projects.
+    use_vim_ft = false,
```
Could we instead make this an option for the `icon` column, like we do for `add_padding`? You can just access it on the `conf` table like `conf.use_slow_filetype_detection` and then add an entry to [this file](https://github.com/stevearc/oil.nvim/blob/685cdb4ffa74473d75a1b97451f8654ceeab0f4a/scripts/generate.py#L140-L152) to generate the documentation for it.

nobe4 committed:
refactor(config): add the key in the generated config

nobe4 committed:
docs: generate the documentation for the new option

nobe4 committed:
fix(column): load file from absolute path

This uses the current oil buffer's path.

nobe4 committed:
fix(lint): format
